File: pkg/taucorder/service/monkey.go
package service

import (
	"context"
	"errors"
	"fmt"

	"connectrpc.com/connect"
	pb "github.com/taubyte/tau/pkg/taucorder/proto/gen/taucorder/v1"
)

func (ms *monkeyService) Get(ctx context.Context, req *connect.Request[pb.GetJobInstanceRequest]) (*connect.Response[pb.Job], error) {
	ni, err := ms.getNode(req.Msg)
	if err != nil {
		return nil, err
	}

	jid := req.Msg.GetId()
	if jid == "" {
		return nil, errors.New("no job id")
	}

	jstatus, err := ni.monkeyClient.Status(jid)
	if err != nil {
		return nil, fmt.Errorf("fetching job failed: %w", err)
	}

	return connect.NewResponse(&pb.Job{
		Id:     jid,
		Status: int32(jstatus.Status),
		Logs: []*pb.JobLog{{
			Cid: jstatus.Logs,
		}}}), nil
}

func (ms *monkeyService) List(ctx context.Context, req *connect.Request[pb.Node], stream *connect.ServerStream[pb.Job]) error {
	ni, err := ms.getNodeById(req.Msg.GetId())
	if err != nil {
		return err
	}

	jids, err := ni.monkeyClient.List()
	if err != nil {
		return fmt.Errorf("list jobs failed: %w", err)
	}

	for _, jid := range jids {
		err = stream.Send(&pb.Job{
			Id: jid,
		})
		if err != nil {
			return err
		}
	}

	return nil
}

File: pkg/taucorder/clients/js/src/Monkey.ts
import { Transport } from "@connectrpc/connect";
import { RPCClient } from "./MonkeyClient";
import { Node, Job } from "../gen/taucorder/v1/common_pb";
import { GetJobInstanceRequest } from "../gen/taucorder/v1/monkey_pb";

export class Monkey {
  private client: RPCClient;
  private node: Node;

  constructor(transport: Transport, node: Node) {
    this.client = new RPCClient(transport);
    this.node = node;
  }

  /**
   * List all jobs
   * @returns AsyncIterable of jobs
   */
  async list(): Promise<AsyncIterable<Job>> {
    return this.client.list(this.node);
  }

  /**
   * Get a specific job instance
   * @param jobId Job ID
   * @returns Job
   */
  async get(jobId: string): Promise<Job> {
    const request = new GetJobInstanceRequest({
      node: this.node,
      id: jobId,
    });
    return this.client.get(request);
  }
} 